File: cpp/src/0082.cpp
/*
[82. Remove Duplicates from Sorted List II (Medium)](https://leetcode.com/problems/remove-duplicates-from-sorted-list-ii/)

@topic: Linked List
@topic: Two Pointers

Constraints:
  - The number of nodes in the list is in the range `[0, 300]`.
  - `-100 <= Node.val <= 100`
  - The list is guaranteed to be **sorted** in ascending order.
*/

#include "structs/LinkedList.hpp"

class Solution {
   public:
    ListNode* deleteDuplicates(ListNode* head) {
        if (!head || !head->next) return head;
        auto fake = ListNode(0, head);
        auto slow = &fake;
        auto fast = head;
        while (fast) {
            while (fast->next && fast->val == fast->next->val) {
                auto prev = fast;
                fast = fast->next;
                delete prev;
            }
            if (slow->next != fast) {
                slow->next = fast->next;
            } else {
                slow = slow->next;
            }
            fast = fast->next;
        }
        return fake.next;
    }

    // ListNode* deleteDuplicates(ListNode* head) {
    //     if (!head || !head->next) return head;
    //     auto fake = ListNode(0, head);
    //     auto curr = &fake;
    //     while (curr->next && curr->next->next) {
    //         auto val = curr->next->val;
    //         if (val != curr->next->next->val) {
    //             curr = curr->next;
    //         } else {
    //             while (curr->next && curr->next->val == val) {
    //                 auto next = curr->next;
    //                 curr->next = next->next;
    //                 delete next;
    //             }
    //         }
    //     }
    //     return fake.next;
    // }
};

#include <cassert>
#include <chrono>
#include <iostream>
#include <vector>

using std::tuple;
using std::vector;
using std::chrono::duration_cast;
using std::chrono::microseconds;
using std::chrono::system_clock;

int main() {
    using namespace LinkedList;

    Solution o;

    vector<tuple<ListNode*, ListNode*>> CASES = {
        {of({1, 2, 3, 3, 4, 4, 5}), of({1, 2, 5})},
        {of({1, 1, 1, 2, 3}), of({2, 3})},
    };

    for (auto& [head, excepted] : CASES) {
        assert(eq(o.deleteDuplicates(head), excepted));
    }

    auto start = system_clock::now();
    for (auto i = 0; i < 100000; ++i) {
        o.deleteDuplicates(of({1, 2, 3, 3, 4, 4, 5}));
        o.deleteDuplicates(of({1, 1, 1, 2, 3}));
    }
    auto end = system_clock::now();
    std::cout << duration_cast<microseconds>(end - start).count() << std::endl;
}

File: java/src/0082.java
import static org.assertj.core.api.Assertions.assertThat;
import static structs.util.LinkedList.of;

import java.time.Instant;
import java.time.temporal.ChronoUnit;

import structs.linkedlist.ListNode;

/**
* Definition for singly-linked list.
* public class ListNode {
*     int val;
*     ListNode next;
*     ListNode() {}
*     ListNode(int val) { this.val = val; }
*     ListNode(int val, ListNode next) { this.val = val; this.next = next; }
* }
*/
class Solution {
    // public ListNode deleteDuplicates(ListNode head) {
    //     if (head == null || head.next == null) {
    //         return head;
    //     }
    //     var fake = new ListNode(0x7fffffff, head);
    //     var curr = fake;
    //     while (curr.next != null && curr.next.next != null) {
    //         if (curr.next.val != curr.next.next.val) {
    //             curr = curr.next;
    //         } else {
    //             var val = curr.next.val;
    //             while (curr.next != null && curr.next.val == val) {
    //                 curr.next = curr.next.next;
    //             }
    //         }
    //     }
    //     return fake.next;
    // }

    public ListNode deleteDuplicates(ListNode head) {
        if (head == null || head.next == null) {
            return head;
        }
        var fake = new ListNode(0, head);
        var slow = fake;
        var fast = head;
        while (fast != null) {
            while (fast.next != null && fast.val == fast.next.val) {
                fast = fast.next;
            }
            if (slow.next != fast) {
                slow.next = fast.next;
            } else {
                slow = slow.next;
            }
            fast = fast.next;
        }
        return fake.next;
    }

    public static void main(String[] args) {
        var o = new Solution();

        assertThat(o.deleteDuplicates(of(1, 2, 3, 3, 4, 4, 5))).isEqualTo(of(1, 2, 5));
        assertThat(o.deleteDuplicates(of(1, 1, 1, 2, 3))).isEqualTo(of(2, 3));

        var start = Instant.now();
        for (var i = 0; i < 10000; i++) {
            o.deleteDuplicates(of(1, 2, 3, 3, 4, 4, 5));
            o.deleteDuplicates(of(1, 1, 1, 2, 3));
        }
        System.out.println(ChronoUnit.MICROS.between(start, Instant.now()));
    }
}
